executor/runtime/docker/docker.go

```diff
@@ -1873,15 +1873,128 @@ func (r *DockerRuntime) startNonMainContainers(ctx context.Context) error {
+			if c == target {
+				return true
+			}
+		}
+	}
+
+	return false
+}
 
+func (r *DockerRuntime) determineContainerOrdering() []string {
 	// First is platform sidecars, then is user-sidecars, then main
 	cNames := r.getPlaformContainerNames()
 	cNames = append(cNames, r.getUserContainerNames()...)
 	cNames = append(cNames, runtimeTypes.MainContainerName)
 
+	origOrder := map[string]int{}
+	for i, cName := range cNames {
+		origOrder[cName] = i
+	}
+
+	thePod, podLock := r.c.Pod()
+	defer podLock.Unlock()
+
+	sort.SliceStable(cNames, func(i, j int) bool {
+		// does i come before j?
+		if inOrderingAnnotation(thePod, cNames[i], pod.AnnotationKeySuffixContainersStartBefore, cNames[j]) {
+			return true
+		}
+
+		// does i come after j?
+		if inOrderingAnnotation(thePod, cNames[i], pod.AnnotationKeySuffixContainersStartAfter, cNames[j]) {
+			return false
+		}
+
+		// does j come before i?
+		if inOrderingAnnotation(thePod, cNames[j], pod.AnnotationKeySuffixContainersStartBefore, cNames[i]) {
+			return false
+		}
+
+		// does j come after i?
+		if inOrderingAnnotation(thePod, cNames[j], pod.AnnotationKeySuffixContainersStartAfter, cNames[i]) {
+			return true
+		}
+
+		// otherwise their order doesn't matter, so let's be stable
+		// w.r.t. the original ordering.
+		return origOrder[cNames[i]] < origOrder[cNames[j]]
+	})
+
+	return cNames
+}
+
+func (r *DockerRuntime) waitForHealthy(ctx context.Context, cName string) error {
+	// XXX: should max tries here? seems like relying on context is enough...
+	for {
+		dockerName := cName
+		if dockerName == runtimeTypes.MainContainerName {
+			dockerName = r.c.TaskID()
+		} else {
+			for _, c := range r.c.ExtraPlatformContainers() {
+				if c.Name == cName {
+					dockerName = c.Status.ContainerID
+				}
+			}
+			for _, c := range r.c.ExtraUserContainers() {
+				if c.Name == cName {
+					dockerName = c.Status.ContainerID
+				}
+			}
+		}
+		inspectOutput, err := r.client.ContainerInspect(ctx, dockerName)
+		if err != nil {
+			return err
+		}
+
+		// sometimes docker gives a NoHealthcheck response, sometimes
+		// it doesn't populate this field at all if there wasn't a
+		// healthcheck.
+		if inspectOutput.State.Health == nil {
+			return nil
+		}
+
+		switch inspectOutput.State.Health.Status {
+		case types.NoHealthcheck:
+			return nil
+		case types.Starting:
+			break
+		case types.Healthy:
+			return nil
+		case types.Unhealthy:
+			msg := "unknown error"
+			exitCode := 255
+			count := len(inspectOutput.State.Health.Log)
+			if count > 0 {
+				// oldest first, so let's pick the last one, i.e. the newest one
+				last := inspectOutput.State.Health.Log[count-1]
+				msg = last.Output
+				exitCode = last.ExitCode
+			}
+			return fmt.Errorf("container %s is unhealthy (last failure %d: %v)", dockerName, exitCode, msg)
+		default:
+			return fmt.Errorf("unknown container health status for %s: %s", dockerName, inspectOutput.State.Health.Status)
+		}
```

Comment:
This is a little weird, because we *already* have code to parse docker events and put them in the pod status?
Can this be reused? Or is it that these goroutines have not started yet I guess?

Reply:
You're talking about the stuff in `statusMonitor`? It looks like we have started those at this point. Not a ton of docs on those events: https://pkg.go.dev/github.com/docker/docker/api/types/events but it might be possible based on https://docs.docker.com/engine/reference/commandline/events/ let me see.

Reply:
Doesn't look like it, we only get `start`, `die`, and `destroy`; I messed around with the [filters](https://pkg.go.dev/github.com/docker/docker/api/types/filters#Args.Add) and couldn't get anything to pop out either :(